### DISCUSSION TITLE
Pybricks for FLL Teams: One-year review

### DISCUSSION THREAD

**Discussion by FLL-Team-24277** [pybricks/support#1329] [General]
How we used pybricks for our FLL team

We are FLL Team 24277 from Norfolk, Virginia, and this year was our first year usiing pybricks. It was our second year using python though, so we had some experience coming in. We use Spike Primes, so I can't really speak for EV3s with pybricks, but it seems that a lot of the functionality is the same. We are a middle school team, which means 12-14 years old.

We just completed our 2023 Season, Masterpiece. We had second place in robot performmance at our local regional event, and did advance to the state tournament, where we actually got first place in robot performance (400 points official, although we did have a 455 unofficial practice run).

Pybricks worked really well for us. We have ten practice robots, and I was worried about bluetooth performance in the classroom where we practice, but that really wasn't a problem. We do see sometimes when we go to upload a program to the hub that sometimes it requires turning the hub off and back on again, but pybricks boots so quickly this is never a problem. The kids do complain that the "timeout" for when the robot turns itself off (presumably to save the battery) seems to be too quick, and we do wish we could configure that.

I think there are still some features that we would like to try out, but never got around to it. For example, we just learned about Stop.NONE and how it can be used to "chain" some movement commands like straight() and curve() and you can really save some time without having to stop between every movement command. I am sure there are some other cool commands that we haven't tried out yet, but we probably should.

Over the summer last year we wrote a BaseRobot class that implements the DriveBase and assigns all of the motors and sensors to the correct ports, so the team won't have to mess with that. It also sets the color sensor parameters and some default speeds that we liked. We wrote wrapper methods for most of the things we will do a lot (like straight and curve) just to make it easier for the programmers. If you decide that you want to use pybricks for your team, I would definitely recommend doing it this way. You can also make a master program that launches your individual mission files as regular python imports, which makes it really easy to keep each mission in its own file.

I was also concerned that importing so many programs would take up too much memory, which would be really problematic and hard to troubleshoot. This turned out to not be a problem for us at all. All of our code totals about 650 lines of code, with probably 80% of that being comments, so call it 200 lines of executable code. Absolutely no problem for pybricks and the hardware to handle.

We also use VS Code and GitHub, which really makes it much easier for the kids to collaborate and share the code that they have written, such as the BaseRobot class. The kids have said that they like programming this way, and I think it is preparing them for advanced robotics programs like FTC and FRC, as well as just being used to using those tools for programming classes, etc.

I'm not sure that pybricks gave us a competitive advantage. I think I could create all of the functionality that pybricks offers within the block coding that comes from Lego. However, sharing that code is very challenging, and that alone sells me on using pybricks.

In all, pybricks seemed very reliable and accurate. Every time we thought we had some problem with pybricks, it turned out to be an error on our part. As for recommending it to other teams, that is hard to say. Since we are talking about 12-14 year-old kids, and coaches that may not have experience with python, I definitely wouldn't recommend it for everyone. If you are a coach that has motivated kids that want to learn python, and you have the means to support them, I would really recommend it.

[Here is our GitHub repository for this season.](https://github.com/FLL-Team-24277/Master-Program-Fall-2023/tree/main)

**Comment by laurensvalk** [comment]
Well done team! Congratulations on your results for this season. We're honored that you used Pybricks, and glad that it worked well for you! :smile:

**Comment by LegoPink** [comment]
Thanks for sharing your experiences with PyBricks in FLL competition! 
I'm a FLL coach myself (since the 2009 Smart Move season) and my teams use PyBricks this season as well. My teammembers are 9-12 year olds (primary school). 
Some kids in the team learned the basics of Python programming (before starting in PyBricks) through Hedy.org. With Hedy, kids gradually learn programming concepts. 

My main reason to switch to PyBricks was frustration with the LEGO software, combined with managed school Chromebooks. Advantages of PyBricks I see:
- Copy / paste code
- Organise code; we have 
-     bot.py file with al the details about motors, sensors etc 
-     do.py with wrappers for straight and turn in Dutch 
-     run1.py run2.py etc (but with more descriptive names then that) with a function def go()
-     start.py that imports all runs and shows a run number on the screen, arrow buttons to toggle 
- Code commenting (with the blockbased programs, kids didn't do commenting. Now it's so easy, they do
- Backups (in our case, to a USB stick. Maybe a shared drive in the cloud next year)
- Gyro in DriveBase
- Ease of coding, monkey see, monkey do

**Reply by boergerj** [reply]
Fellow FLL Coach here,

My school system also uses Chromebooks. How are you using PyBricks? How would you integrate this with a shared drive for students to share code?

I'm trying to find the best setup for the students to use so it's easy to edit, have multiple files, and to share their code. Is there a way to use github easily with Chromebooks and PyBricks?

**Reply by LegoPink** [reply]
We went old school: USB Stick. After each session, teammembers would backup to stick, and upload the latest backup from stick the next session. Since they only have the one robot, only one set of kids would be programming at any given time.
I could have had them use a shared drive, but at events it can be a struggle to get onto the internet and into the cloud.

**Comment by CCz24** [comment]
Hi! I'm looking at using pybricks this year with my  FIRST team. The main reason is that I think Lego maybe didn't set up the gyro in the Spike Prime correctly-- it is glitchy and jumps around, making the gyro unusable. We did a lot of testing last year, when our Spike Prime robots were new. And the internal gyro doesn't work at all using Lego programming. (Two complete circles with a series of  90 degree turns should wind up in same position, it was very much off) But... It looks like Pybricks figured it out? Maybe?? Can anyone confirm the gyro will make accurate turns with the gyro using pybricks? Thanks so much. That is really the only reason we might use it. We used Python last year in LEGO's setup and everything worked well and copied well except the gyro problem. This would make a world of difference if it works!

**Reply by FLL-Team-24277** [reply]
We were very pleased with the gyro performance last year, and we are using it this year too. It's not perfect, but it's not bad.

**Reply by laurensvalk** [reply]
Here is an indication of how [accurate that can get with Pybricks](https://www.youtube.com/watch?v=tzoGeR_OFhQ). :smile: 

The gyro value doesn't reset itself at +/- 180, so working with the gyro is going to be a bit easier in Pybricks.

Still, keep in mind that the gyro is not a magic bullet. It is not a compass, so it has no way of truly measuring what 90-degrees really is. Each gyro will be a bit off (e.g. measure 87 or 92 degrees) but do so consistently. If you don't account for this and make many turns _in the same direction_ this is going to add up. You can account for this in your code, and we're introducing a feature to make this calibration step a bit easier.

**Reply by CCz24** [reply]
Laurensvallk, WOW. That video is amazing! That's what we needed so much last year. We wound up not using the gyros at all because we had such issues with them, they even made the bot spin around a few instances! We have the numbers printed (degree turns) and I'll see if I can dig up that info again. A negative value is thrown in at random and really messes things up. When we removed the gyro we did use a correction factor for turns that worked very well, but yes it seemed just lucky when runs actually worked. Two years ago we used the EV3 and had the gyro working so well that if the robot ran over a Lego, and got off course, it would self correct. We don't really need that,  but wow was that fun! I am really impressed with your video, and 24277 thanks for your note of continuation, I hope to convince the team to give Pybricks a try.

**Reply by laurensvalk** [reply]
>  had the gyro working so well that if the robot ran over a Lego, and got off course, it would self correct

Right, that's even a builtin function with Pybricks. If you use `straight()` with the gyro enabled, it's going to self-correct.

> I am really impressed with your video

We should probably make more of these. I like [this video](https://www.youtube.com/shorts/OWmU6iIPLU0) from coach @ggramlich a lot (reposted on our channel with English subs with permission. Original is [here](https://www.youtube.com/@ProgrammierenFuerKids).)

**Reply by CCz24** [reply]
Yes! Consistent positioning is the goal for sure. (after many turns and movements, of course!) I uploaded our testing code and two results for examples. The negative values, and the general jumping around. Just... wow. No wonder our turns were never quite the same. I'm looking for a video of a spin, or a video of this testing. I think the kids might have it, I'll ask them. But I think you know our issue! And Pybricks seems like a logical leap. Now to convince the teams! Many thanks for the videos and notes. https://github.com/CCz24/Gyro-trial-2023

**Comment by CCz24** [comment]
We have Pybricks working! It took some time to try to figure out how to
"remove" the block code (we used Python last year in Lego's format) and
woowoowww that is neat. The gyro WORKS!! This is amazing!! Thank you for
getting us to try Pybricks!!
A question: the motor.stop is supposed to let the motors go... make them
spin freely? We would like to figure this out so we can just move the bot
around freely to figure out our missions. Is this possible? Or do we just
have to make a similar model, sans motors? (we would rather just switch the
motors to spin freely!) Thanks for any help there! Maybe it isn't possible?
-Carrie

**Reply by laurensvalk** [reply]
I'm glad to hear it! :smile: 

> A question: the motor.stop is supposed to let the motors go... make them spin freely

Compared to the motors holding position, then yes, `stop()` makes them spin "freely", but you'll still need to push them a bit to overcome friction internal to the motor. If the wheels have enough grip, usually pressing gently on the robot will drag the wheels along.

**Comment by CCz24** [comment]
Yes, I think we might need to just make a "free spinning" model. It's only
in one direction, and for a short time it seems.
Thanks again!

**Comment by CCz24** [comment]
Hello Laurens!

We have started PyBricks block code with our FLL team of 6. We loooove the
gyro functionality!! This is what we have been looking for ever since w
started using the Spike Prime! Thank you so very much for getting us
started with PyBricks.
We figured out how to start with color blocks in Python via a menu. But we
haven't been able to figure that out yet in block code. And guess what, our
team chose to use block code. Is it possible to begin a run with color? I
saw you have this:
https://pybricks.com/project/spike-hub-menu/#multiple-programs-on-other-hubs

And it seems you have to switch to Python at one point, and I'm wondering
if a color start menu is the same situation? Is it even possible to do
menuing with color in a block code program?

Then the other thing, my team is wanting to work on both sides of the table
at the same time. They have duplicated their bot. Is it possible for them
to run and make changes to the same single program at the same time? The
saving seems to happen on the computer (we have two computers) and they
struggle with the saving bit. (I'm not sure at all how to help there!) I am
used to saying python code to a Google drive. We can't do that with block
code though, correct?

We have forgotten computers several times at competition in the past two
years, so it makes sense to keep a current code online somewhere. Plus,
working with two computers, it makes sense to take the updated program at
the start, on either device. Is there instructions for saving?

Maybe we need to do two completely different programs for practice, then
combine them before competition. Not ideal, but I don't think we can create
independent runs like you can with Lego, correct?

The start up of PyBricks has been a challenge, but they are excited. Thank
you for any direction here!

Best wishes,
Carrie Czerwonka and the Lobster Mafia
***@***.***

**Reply by laurensvalk** [reply]
> We have started PyBricks block code with our FLL team of 6. We loooove the
gyro functionality!! This is what we have been looking for ever since w
started using the Spike Prime! Thank you so very much for getting us
started with PyBricks.

Thank you! We are glad to hear it.


> And it seems you have to switch to Python at one point

The [technique you referenced](https://pybricks.com/project/spike-hub-menu) only uses Python for the main menu.

> We figured out how to start with color blocks in Python via a menu. But we
haven't been able to figure that out yet in block code. 

What do you mean with "color blocks"? Do you mean something different than block coding?

Do you mean you would like to start different parts of your code using the color sensor?

> I am used to saying python code to a Google drive. We can't do that with block
code though, correct?

Block programs and Python programs are treated the same. You can even download all projects at once. See also: https://pybricks.com/learn/getting-started/pybricks-environment/#managing-program-files

There is no "Google Drive"-style automatic synchronization if that's what you mean.

> Maybe we need to do two completely different programs for practice, then
combine them before competition. Not ideal, but I don't think we can create
independent runs like you can with Lego, correct?

I'm not entirely sure what you mean with independent runs. If you mean to store multiple programs and access them with a menu, you can use the Python technique referenced above.

We understand that this is can be difficult for beginners, so we are also working on a way to simply upload several separate programs, without having to make your own menu. This may come in a future release.
